File: Leetcode/Leetcode-DSA-course/Hashing/3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        Map<Character, Integer> counts = new HashMap<>();
        int left = 0, ans = 0;
        for (int right = 0; right < s.length(); right++) {
            char r = s.charAt(right);
            counts.put(r, counts.getOrDefault(r, 0) + 1);
            while (counts.get(r) > 1) {
                char l = s.charAt(left);
                counts.put(l, counts.get(l) - 1);
                left++;
            }
            ans = Math.max(ans, right - left + 1);
        }
        return ans;
    }
}

File: Leetcode/Leetcode-DSA-course/Hashing/3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int left = 0, ans = 0;
        unordered_map<char, int> counts;
        for (int right = 0; right < s.length(); right++) {
            counts[s[right]]++;
            while (counts[s[right]] > 1) {
                counts[s[left]]--;
                left++;
            }
            ans = max(ans, right - left + 1);
        }

        return ans;
    }
};